rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -5233,6 +5233,102 @@ extern "C" rmw_ret_t rmw_count_subscribers(
   return common_context->graph_cache.get_reader_count(mangled_topic_name, count);
 }
 
+extern "C" rmw_ret_t rmw_count_clients(
+  const rmw_node_t * node, const char * service_name,
+  size_t * count)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_name, RMW_RET_INVALID_ARGUMENT);
+  int validation_result = RMW_TOPIC_VALID;
+  rmw_ret_t ret = rmw_validate_full_topic_name(service_name, &validation_result, nullptr);
+  if (RMW_RET_OK != ret) {
+    return ret;
+  }
+  if (RMW_TOPIC_VALID != validation_result) {
+    const char * reason = rmw_full_topic_name_validation_result_string(validation_result);
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("service_name argument is invalid: %s", reason);
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  RMW_CHECK_ARGUMENT_FOR_NULL(count, RMW_RET_INVALID_ARGUMENT);
+  auto common_context = &node->context->impl->common;
+  const std::string mangled_rq_service_name =
+    make_fqtopic(ROS_SERVICE_REQUESTER_PREFIX, service_name, "Request", false);
+  const std::string mangled_rp_service_name =
+    make_fqtopic(ROS_SERVICE_RESPONSE_PREFIX, service_name, "Reply", false);
+
+  size_t number_of_request_publishers = 0;
+  ret = common_context->graph_cache.get_writer_count(
+    mangled_rq_service_name,
+    &number_of_request_publishers);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  size_t number_of_response_subscribers = 0;
+  ret = common_context->graph_cache.get_reader_count(
+    mangled_rp_service_name,
+    &number_of_response_subscribers);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  if (number_of_request_publishers != number_of_response_subscribers) {
+    return RMW_RET_ERROR;
+  }
+  return common_context->graph_cache.get_reader_count(mangled_rp_service_name, count);
```

Comment:
If you consider it an error when the number of publisher & the number of subscribers don't match, then arguably you should here return `number_of_response_subscribers`, because the graph cache may already have changed here.